File: src/main/java/injea/knwremodel/Notice/NoticeController.java
package injea.knwremodel.Notice;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import injea.knwremodel.Like.LikeService;

import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/notice")
@RequiredArgsConstructor
public class NoticeController {
    private final NoticeService noticeS;
    private final LikeService likeS;
    
    
    @PostMapping("/requestPage")
    public ResponseEntity requestPage(@RequestBody NoticeDTO.requestPage dto) {
        PageRequest paging = PageRequest.of(dto.getPage().intValue(), dto.getPerPage().intValue(),Sort.Direction.DESC, "id");
        Page<Notice> notices = noticeS.search(dto.getMajor(), dto.getType(), dto.getKeyword(), paging);
        Page<NoticeDTO.responsePage>result = notices.map(notice -> new NoticeDTO.responsePage(likeS, notice));
        return ResponseEntity.ok(result);
    }

    @PostMapping("/requestbody")
    public ResponseEntity requestbody(@RequestBody NoticeDTO.requestbody dto) {
        Notice notice = noticeS.findById(dto.getDbid());
        return ResponseEntity.ok(new NoticeDTO.responsebody(likeS, notice));
    }

    @PostMapping("/toplike")
    public ResponseEntity getTopLikeByMajor(@RequestBody NoticeDTO.toplike dto) {
        List<Notice> topNotices = noticeS.findTopLike(dto.getMajor(), PageRequest.of(0, dto.getSize(), Sort.Direction.DESC, "likeCount"));
        List<NoticeDTO.responsebody> result = topNotices.stream().map(notice -> new NoticeDTO.responsebody(likeS, notice)).collect(Collectors.toList());
        return ResponseEntity.ok(result);
    }

    @PostMapping("/topView")
    public ResponseEntity getTopView(@RequestBody NoticeDTO.topview dto) {
        List<Notice> topNotices = noticeS.findTopView(PageRequest.of(0, dto.getSize(), Sort.Direction.DESC, "view"));
        List<NoticeDTO.responsebody> result = topNotices.stream().map(notice -> new NoticeDTO.responsebody(likeS, notice)).collect(Collectors.toList());
        return ResponseEntity.ok(result);
    }
}

File: src/main/kotlin/injea/knwremodel/Notice/NoticeController.kt
package injea.knwremodel.Notice

import injea.knwremodel.Like.LikeService
import injea.knwremodel.Notice.NoticeDTO.*
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.stream.Collectors

@RestController
@RequestMapping("/notice")
class NoticeController(private val noticeS: NoticeService, private val likeS: LikeService) {
    @PostMapping("/requestPage")
    fun requestPage(@RequestBody dto: requestPage): ResponseEntity<*> {
        val paging = PageRequest.of(dto.page.toInt(), dto.perPage.toInt(), Sort.Direction.DESC, "regdate")
        val notices = noticeS.search(dto.major, dto.type, dto.keyword, paging)
        val result = notices?.map { notice: Notice ->
            responsePage(
                likeS, notice
            )
        }
        return ResponseEntity.ok(result)
    }

    @PostMapping("/requestbody")
    fun requestbody(@RequestBody dto: requestbody): ResponseEntity<*> {
        val notice = noticeS.findById(dto.id)
        return ResponseEntity.ok(responsebody(likeS, notice))
    }

    @PostMapping("/toplike")
    fun getTopLikeByMajor(@RequestBody dto: toplike): ResponseEntity<*> {
        val topNotices = noticeS.findTopLike(dto.major, PageRequest.of(0, dto.size, Sort.Direction.DESC, "likeCount"))
        val result = topNotices?.map { notice: Notice ->
            responsePage(
                likeS, notice
            )
        }
        return ResponseEntity.ok(result)
    }

    @PostMapping("/topView")
    fun getTopView(@RequestBody dto: topview): ResponseEntity<*> {
        val topNotices = noticeS.findTopView(PageRequest.of(0, dto.size, Sort.Direction.DESC, "view"))
        val result = topNotices?.map { notice: Notice ->
            responsePage(
                likeS, notice
            )
        }
        return ResponseEntity.ok(result)
    }
}